app/ethos/views.py

```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+'''
+    Copyright (C) 2017 Gitcoin Core
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published
+    by the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+'''
+
+import json
+
+from django.conf import settings
+from django.http import Http404, JsonResponse
+from django.template.response import TemplateResponse
+from django.utils import timezone
+from django.utils.translation import gettext_lazy as _
+from django.views.decorators.csrf import csrf_exempt
+
+import twitter
+from ethos.models import Hop, ShortCode
+from gas.utils import recommend_min_gas_price_to_confirm_in_time
+from ratelimit.decorators import ratelimit
+from retail.helpers import get_ip
+from web3 import HTTPProvider, Web3
+
+w3 = Web3(HTTPProvider(settings.WEB3_HTTP_PROVIDER))
+
+
+@csrf_exempt
+@ratelimit(key='ip', rate='5/m', method=ratelimit.UNSAFE, block=True)
+def redeem_coin(request, shortcode):
+    print(shortcode)
+
+    if request.body:
+        status = 'OK'
+
+        body_unicode = request.body.decode('utf-8')
+        body = json.loads(body_unicode)
+
+        try:
+            twitter_api = get_twitter_api()
+
+            if not twitter_api:
+                raise Exception(_('Error while setting up Twitter API'))
+
+            address = body['address']
+            username = body['username']
+
+            twitter_api.GetUser(screen_name=username)
```

Comment:
Should we assign the return value here or drop this?

Reply:
We're validating whether the given input of Twitter username is valid or not. If it's not valid, this throws an exception which is being handled below.

Reply:
@eswarasai  Yeah, I gotcha on that. I'm just wondering if there would be a better way to handle validating the user. The user could enter any twitter handle that is valid and this will be successful.  Do we want to validate the user truly is the twitter user they are claiming to be? If not, we're performing an API call to check that an arbitrary user exists in the twitterverse.

Reply:
@mbeacom -- I'm not really sure on how we can truly validate whether the provided Twitter username belongs to the user who's redeeming the coin

Reply:
@eswarasai My concerns with the current implementation are:

- We're checking that the twitter handle exists on twitter, but not validating the user entering the twitter handle actually owns that twitter account.
- We are tweeting the same message for every redemption as the EthOS twitter account and not `@` mentioning them.
- Tweeting as the EthOS account (even with only 50 coins) could end with the EthOS account hitting it's throttling limit since we're not using the user's twitter to tweet out about the coin redemption.
- If we authed the user with twitter, we could tweet out AS them, `@` mention the EthOS twitter handle, and tag it with the appropriate `#` and periodically tweet out as the EthOS account with the force directed graph.
- If we decide to go the route of not authing the user and tweeting as them, we might as well eliminate the check on the twitter handle entirely since it's not doing anything other than checking that the user exists.

Over all, everything looks good.  I just think we might want to approach the twitter interaction a little differently.